### DISCUSSION TITLE
Inquiry on Hyperparameter Importance in Optuna: Impotance values and order are changing

### DISCUSSION THREAD

**Discussion by GhsJahanmir** [optuna/optuna#5680] [Q&A] [answered]
Dear All,

I am currently working on a hyperparameter optimization task for a neural network model using **Optuna**. I have selected nearly 10 parameters for tuning. After conducting the optimization, I run the following command to obtain the importance values of the tested hyperparameters:

**optuna.importance.get_param_importances(study)**

However, I've noticed that each time I execute the script (Not optimization process), the order and values of the importance values change. Could you please help me understand the reasons behind this variability? Additionally, what would be the best approach to report the most important hyperparameters for further fine-tuning?

Thank you in advance for your assistance!
Best Regards,
Jahanmir

**Comment by nzw0301** [comment, marked as answer]
> Could you please help me understand the reasons behind this variability?

I think the default importance evaluator, https://optuna.readthedocs.io/en/stable/reference/generated/optuna.importance.FanovaImportanceEvaluator.html#optuna.importance.FanovaImportanceEvaluator, has some randomness that can be fixed by specifying `seed` argument of the evaluator.

**Comment by nzw0301** [comment]
> what would be the best approach to report the most important hyperparameters for further fine-tuning?

Could you explain the details of this question? What do you mean further fine-tuning?

**Reply by GhsJahanmir** [reply]
> > what would be the best approach to report the most important hyperparameters for further fine-tuning?
> 
> Could you explain the details of this question? What do you mean further fine-tuning?


Thanks for your reply. I plan to get the most important parameters first from initial grids, Freeze the least important parameter values and then rerun the experiment with a finer grid (only tune the most important hyperparameters) to get more robust estimate of hyperparameters. 

Does this approach make sense?

I used the default sampler and trial number around 200 to 300.

**Comment by nabenabe0928** [comment]
As mentioned by @nzw0301 , the default importance evaluator involves randomness.
You can either fix the seed or alternatively use a deterministic evaluator such as [PED-ANOVA](https://optuna.readthedocs.io/en/latest/reference/generated/optuna.importance.PedAnovaImportanceEvaluator.html).

**Reply by GhsJahanmir** [reply]
> As mentioned by @nzw0301 , the default importance evaluator involves randomness. You can either fix the seed or alternatively use a deterministic evaluator such as [PED-ANOVA](https://optuna.readthedocs.io/en/latest/reference/generated/optuna.importance.PedAnovaImportanceEvaluator.html).


Thank you for your suggestion.

does fixing the TPE sampler resolving the problem or I need to fix the seed of evaluator too?

my intention is to get the most important parameter first and then keep others constants and rerun the experiment using a finer grid of the the most influential parameters. considering my objective, what would you suggest me to do that?

any comment on the number of trials?

**Reply by nabenabe0928** [reply]
So here are some points:
1. Parameter importance changes based on observations, so if you would like to reproduce the importance order, the seed of a sampler must be fixed
2. Once a fixed study is given, PED-ANOVA always returns the same results while fANOVA needs a fixed seed.

I would personally recommend PED-ANOVA due to its interpretability. 
Note that PED-ANOVA tells you which parameters were important to achieve the top 10% in a study.
Given the context, we need as many trials as possible, but let's say minimum 20 Trials.

**Reply by nabenabe0928** [reply]
Just a comment, but how would you pick constant values?

My recommendation would be actually to fix the most influential Parameter to the one in the best trial and check the other Parameters.

That's because the variation in the most influential Parameter easily hides the influence from the other parameters.
But most importantly, you cannot really know what constant values to pick for the other parameters in your methodology.

**Reply by GhsJahanmir** [reply]
> Just a comment, but how would you pick constant values?
> 
> My recommendation would be actually to fix the most influential Parameter to the one in the best trial and check the other Parameters.
> 
> That's because the variation in the most influential Parameter easily hides the influence from the other parameters. But most importantly, you cannot really know what constant values to pick for the other parameters in your methodology.

Thank you for your comment.
That is a good point.

I am still uncertain about the basis of the procedure. Is it really logical to use the importance valued  as a guide to choose the parameters? because if we check the range of parameters  used by trials, (plot one by one against objective value) we can see that some of them are not uniformly distributed  and aggregate more around some specific range or values ( for categoriacal variables like optimizer ,,.., one of the optimizers is tested more than others).

I plan to use the individual plot of least important hyperparameters against the objective value and choose based on the range that can give me higher objective functions. For example for number of units, choose the lowest range that can give objective values close enough to the best trial objective values. This approach is perfomed by some literature studies. They fixed some as they do not have much influence in their result. then they tune the two remaining influential parameters in a simple grid search (for them learning rate and number of units).

 But your point is a also valid and it is along the "Andrew Ng" recomendation in the deep learning course. Then I may proceed with that, fixed the important ones first and tune the others after that.

**Reply by GhsJahanmir** [reply]
> So here are some points:
> 
> 1. Parameter importance changes based on observations, so if you would like to reproduce the importance order, the seed of a sampler must be fixed
> 2. Once a fixed study is given, PED-ANOVA always returns the same results while fANOVA needs a fixed seed.
> 
> I would personally recommend PED-ANOVA due to its interpretability. Note that PED-ANOVA tells you which parameters were important to achieve the top 10% in a study. Given the context, we need as many trials as possible, but let's say minimum 20 Trials.


Instead of fixing the seed, can we do multiple runs and use voting or aggregation of the result to get the importance values?


Because I think if we fix the seed (even in sampler and Test_train splitting,...), the generalization ability of the final model would be low. 

This is a general question I have. should we really fix the seed (for numpy and tensor flow on the top ) and for TPE sampler when developing models for reaseach works (for journal publications)?
